mypy/checker.py

```diff
@@ -984,7 +984,12 @@ def is_unannotated_any(t: Type) -> bool:
         check_incomplete_defs = self.options.disallow_incomplete_defs and has_explicit_annotation
         if show_untyped and (self.options.disallow_untyped_defs or check_incomplete_defs):
             if fdef.type is None and self.options.disallow_untyped_defs:
-                self.fail(message_registry.FUNCTION_TYPE_EXPECTED, fdef)
+                if (not len(fdef.arguments) or
+                        (len(fdef.arg_names) == 1 and 'self' in fdef.arg_names)):
```

Comment:
`fdef.arg_names[0] == 'self'` would be more straightforward. Also maybe you should check for both `self` and `cls`.